src/psyclone/domain/gocean/nodes/gocean_extract_node.py

```diff
@@ -141,66 +142,111 @@ def create_driver(self, input_list, output_list):
+            # Support GOcean properties, which are accessed via a
+            # derived type (e.g. 'fld%grid%dx'). In this stand-alone
+            # driver we don't have the derived type, instead we create
+            # variable based on the field in the derived type ('dx'
+            # in the example above), and pass this variable to the
+            # instrumented code.
+            last_percent = var_name.rfind("%")
+            if last_percent > -1:
+                # Strip off the derived type, and only leave the last
+                # field, which is used as the local variable name.
+                local_name = var_name[last_percent+1:]
+            else:
+                # No derived type, so we can just use the
+                # variable name directly in the driver
+                local_name = var_name
+            unique_local_name = sym_table.new_symbol_name(local_name)
+            rename_variable[local_name] = unique_local_name
+            sym_table.add(Symbol(unique_local_name))
+            local_name = unique_local_name
+
+            # TODO: #644 - we need to identify arrays!!
             # Any variable used needs to be defined. We also need
             # to handle the kind property better and not rely on
             # a hard-coded value.
-            decl = DeclGen(prog, "real", [var_name], kind="8",
+            decl = DeclGen(prog, "real", [local_name], kind="8",
                            dimension=":,:", allocatable=True)
             prog.add(decl)
             is_input = var_name in input_list
             is_output = var_name in output_list
 
             if is_input and not is_output:
                 # We only need the pre-variable, and we can read
-                # it from the file (and allocate its size)
+                # it from the file (which also allocate space for it)
                 call = CallGen(prog,
-                               "psy_data%ReadVariable(\"{0}\", {0})"
-                               .format(var_name))
+                               "{0}%ReadVariable(\"{1}\", {2})"
+                               .format(psy_data, var_name, local_name))
                 prog.add(call)
             elif is_input:
                 # Now must be input and output:
-                # First read the pre-variable
+                # First read the pre-variable (which is also allocated):
                 call = CallGen(prog,
-                               "psy_data%ReadVariable(\"{0}\", {0})"
-                               .format(var_name))
+                               "{0}%ReadVariable(\"{1}\", {2})"
+                               .format(psy_data, var_name, local_name))
                 prog.add(call)
                 # Then declare the post variable, and and read its values
                 # (ReadVariable will also allocate it)
-                decl = DeclGen(prog, "real", [var_name+post_suffix],
+                decl = DeclGen(prog, "real", [local_name+post_suffix],
```

Comment:
Strictly speaking, we should use the SymbolTable for the name of the post variable too (in the unlikely case that `local_name+post_suffix` clashes with something).

Reply:
Gee, how could I miss that. This actually showed me that there is a bit of trouble still around. I can easily add code to abort if the declared name already exists, that's easy.

But when creating a test case, I realised that I could actually create a duplicated key in the netcdf file:
```
  call invoke( compute_kernel(out_fld, in_out_fld, out_fld_post, dx))
```
If ``out_fld`` is written, it will create an entry in the netcdf file ``out_fld_post`` (to store the output value). But at the same time if the field ``out_fld_post`` is an input parameter, it is stored as ``out_fld_post``. I need to think about this - either we just don't allow a name that clashes with a "..._post" name, or we need to detect the name clash by renaming one of the keys of the netcdf file, then making the driver-creation function aware of this rename as well 😢

Reply:
OK, in short the problem: we have an output variable 'a', and an input variable 'a_post' --> both will create a key in the netcdf file with the name 'a_post' (which then in turn also creates duplicated variable names in the driver).

I came up with the following (neat??) solution: I added a callback from the ExtractNode after all variables have been modified. This callback can be overwritten, so in GOCeanExtractNode I check if any output-variable when added "_post" is an input variable. If so, I modify the suffix ('post' --> 'postt'), then I repeat till I have a suffix that creates no duplicated keys (appending more and more 't's).

This solves two problems:
1) It makes sure all keys in the netcdf file are unique.
2) It also implies that the variable names created in the driver are unique (since the driver uses the suffix determined above, it will use "_postt" if required).

I call it neat, since during this call we could also remove loop-variables (which we know in the domain-specific node: gocean knows the name of the loop variables, to they can be removed from the list of variables to be written) - though I will do this as a separate PR later.

I am not touching LFRic at this stage, so I have left this all in GOceanExtractNode. If we need the same for LFRic, the code can be moved into ExtractNode.

Reply:
Oh, and yes I still add the new name to the symbol table, so we will get an exception if something should go wrong.

Reply:
I'm happy with that although, as I've commented somewhere else, I think it would be nicer to append an integer rather than 't' :-)

Reply:
Done, including integer :)